### PR TITLE
File::findFiles - Save 200 million nanoseconds

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------

This is a small performance tweak for `findFiles()` which converts `readdir()`+`is_dir()` to `glob(ONLYDIRS)`.

I benchmarked against an arbitrary sample of unit tests (*`tests/phpunit/api/v3/AddressTest.php` with `--filter testCreate`*), and the patch saved ~200 million nanoseconds (~0.2sec) on my workstation.

ping @demeritcowboy 

Before
----------------------------------------

More time:

Screenshot from 2022-09-08 13-59-34

After
----------------------------------------

Less time:

Screenshot from 2022-09-08 14-03-38

Technical Details
----------------------------------------

(*Note: Before profile, I disabled the `DebugSubscriber`.*)

I profiled the following command:

```
XDEBUG_MODE=profile CIVICRM_UF=UnitTests phpunit8 tests/phpunit/api/v3/AddressTest.php --stop-on-failure --filter testCreate
```

This command was repeated a few times with different codebases:

```
TRIAL/FILE NAME                 CRM_Utils_File::findFiles       CODEBASE
cachegrind.out-none             27M ns                          (before #23854)
cachegrind.out-orig             804M ns                         (after #23854)
cachegrind.out-patch1           326M ns                         (after #23854 plus this patch)
cachegrind.out-orig-b           364M ns                         (after #23854)
cachegrind.out-patch1-b         136M ns                         (after #23854 plus this patch)
```

In the first trial (`after` vs `patch1`), it saved ~450M ns. In the second trial (`after-b` vs `patch1-b`), it saved ~230M ns. One might need more trials to establish statistical significance, but it seems consistently better so far, and the change is small.

This may be a bit academic, but I *believe* the difference turns on `is_dir()` -- with the old patch, the call to `is_dir()` would hydrate [PHP's `stat` cache](https://www.php.net/manual/en/function.clearstatcache.php) for every inode in the tree. But `glob()` doesn't affect the `stat` cache.

### CODE REVIEW

**Comment by agileware-justin** [comment]
Does this findfiles function scan the entire webroot directory?

**Comment by totten** [comment]
Ran a few more trials:

* With 2 trials of the full `tests/phpunit/api/v3/AddressTest.php`, it saved almost 400M ns. (635-644 vs 242-259).
* With 2 trials of the `tests/phpunit/CRM/Extension/`, it saved 200M ns (244-296 vs 33-34).

> (@agileware-justin) Does this findfiles function scan the entire webroot directory?

Only if you tell it to. :smile: 

```
CRM_Utils_File::findFiles(Civi::paths()->getPath('[cms.root]/'), '*');
```

Off the top of my head, can't think of anything which is that open-ended.

**Comment by demeritcowboy** [comment]
Although @agileware-justin on older installs it probably scans too much - depends if you added this to civicrm.settings.php: https://github.com/civicrm/civicrm-core/blob/0f5c0fcbeccb1412394848a913621bbd40ac8c0a/templates/CRM/common/civicrm.settings.php.template#L547

For new-ish installs that's the default.

I recommend adding `|vendor/` in there too except for drupal 8.

**Comment by agileware-justin** [comment]
"By default this will search every directory under the docroot, except those that begin with a dot"
https://docs.civicrm.org/sysadmin/en/latest/setup/optimizations/#exclude-dirs-that-do-not-need-to-be-scanned

Thanks @totten @demeritcowboy for the quick replies.

Can anyone ELI5 why scanning the docroot was ever required?Surely the scope of the findfiles should just to strictly limited to CiviCRM directories which have been declared and nothing else.

If a website has millions of files in the file system, CiviCRM is going to scan them all? Why?

Would it not have been better to have added "inclusion" rule rather than an "exception" rule to civicrm.settings.php - since as you say this problem still occurs on older installs. But don't older installs make up the majority of the install base?

Apologies for hijacking, but the performance issues caused by findfiles has been on our bug list for a while and it seems relevant here since you are talking about fixing performance of this very function.

**Comment by totten** [comment]
@agileware-justin A couple things:

To clarify origin, this is an offshoot from @demeritcowboy's investigation into test performance, wherein we found that #23854 correlates to a slow-down. Interestingly, the slowdown does not appear to be in `ClassScanner` per se, but that's leaves an open question for why/how it changed. Current speculation: as a side-effect of a bugfix in 23854, the test-suites do more extension-scanning (*ie searching for `info.xml` files*).

> Can anyone ELI5 why scanning the docroot was ever required? Surely the scope of the findfiles should just to strictly limited to CiviCRM directories which have been declared and nothing else.

AFAIK, that documentation is misleading. It's never searched the docroot. The biggest search that it does is probably for `info.xml` files, which hits: the `[civicrm.root]`, the custom extensions dir (`[civicrm.files]/ext`), and (*if defined*) the CMS `vendor` folder. I'm guessing the last one is the point of contention. Basically, this turns on how much flexibility you give developers+integrators:

* If you include the CMS `vendor`, then it's a small step (*from the client pov*) to using `composer require` to download extension code.
* If you do sub-tree searches, then integrators and developers can organize their extensions into folders.

That one could probably be optimized more (eg require an opt-in for `vendor` search -- although I'd be tempted to do an option like `$maxDepth=3` for `info.xml` searches).

**Comment by eileenmcnaughton** [comment]
image

**Comment by demeritcowboy** [comment]
* General standards
    * [`r-explain`] __PASS__ 🕐 
    * [`r-user`] __PASS__ 🕑 
    * [`r-doc`] __PASS__ 🕒
      * As noted above maybe the docs could use a tweak but is a bit separate from the PR.
    * [`r-run`] __PASS__ 🕓 
      * glob seems to handle everything the original did: excludes dot-dirs, sort of handles is_readable the same way (it tries to descend but silently returns array())
      * output is the same before and after, misc CLI and UI.
      * and seems to handle the civicrm.settings.php setting the same way
      * oh, also it's faster
* Developer standards
    * [`r-tech`] __PASS__ 🕕 
    * [`r-code`] __PASS__ 🕗  
    * [`r-maint`] __PASS__ 🕙 
    * [`r-test`] __PASS__ 🕛

**Comment by eileenmcnaughton** [comment]
image


How does that test run time compare?

**Comment by totten** [comment]
> About 15 min faster. It doesn't fix the cause of the slowdown, just is a bit better.

Actually... `test-1`'s average time for a full/recent PR test is ~2h 48m... so that's ~1.44 trillion nanoseconds faster...

**Comment by eileenmcnaughton** [comment]
@totten my understanding is that issues with file cataloguing is a test only issue because live sites cache that data & it's only on test that it is relentlessly flushed / re-indexed?

**Comment by totten** [comment]
@eileenmcnaughton - Agree. I also expect the impact of this patch to be much less significant in production. (No expected impact on regular pages; maybe a tiny/sub-second impact on flushes.)

**Comment by eileenmcnaughton** [comment]
@totten what worked for us in production recently was reducing calls to serialize large arrays - the count is out on the fastest method to serialize them but doing it less gives big wins... In our case it was reducing fixing a situation where it bypassed the FastArrayDecorator - we ALSO got a measurable benefit eliminating cache misses on small arrays so bigger is slower but if you are serializing / unserialized when (eg.) you access  `\Civi\Core\ClassScanner::cache('index')`  - then that will impact performance.

Last year we  eliminated 50 queries per queue item (contact + contribution import) & our processing rate was unchanged,  but fixing one php-cache bypass per queue item made a massive difference. We determined it was the serialization, not the `Redis` interaction that was the issue . There are a couple of issues still open but we got such a big improvement from https://github.com/civicrm/civicrm-core/pull/24156 & to a lesser degree https://github.com/civicrm/civicrm-core/pull/24294 that we haven't been digging into those

Still open  https://lab.civicrm.org/dev/core/-/issues/3813 & https://lab.civicrm.org/dev/core/-/issues/3814

**Comment by totten** [comment]
Thanks, I hadn't seen some of those links @eileenmcnaughton. I added comments on a few of them.

> we ALSO got a measurable benefit eliminating cache misses on small arrays so bigger is slower but if you are serializing / unserialized when (eg.) you access \Civi\Core\ClassScanner::cache('index') - then that will impact performance.

I feel like this is very interesting, but I can't quite parse it. 

Would it help to discuss the `index` cache more concretely? We could verbalize the expected shape/usage-pattern for data in the `index`? (*ie what records are currently in there, how big they might get, how often they'd be read or written*)

**Comment by eileenmcnaughton** [comment]
@totten I suspect some low level tests on performance of `json_encode` & `serialize` might give some clues - but I found htat people on the interweb had done that - with differing results depending on php version - however what I didn't see much of was how much the size of the array mattered.
